bpf: add -nostdinc and a few more misc compilation options

borkmann opened the pull request:
See commit msg.

borkmann committed:
bpf, misc: fix shadow declarations and declarations after statements

Few random minor changes over the place to fix up warnings in order
to enable stricter compile checks in the subsequent patch. This also
fixes up some whitespace oddities in nodeport code that slipped into
our tree.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: simplify session affinity api and reduce ifdefs in socklb

Fetch the get_netns_cookie() only once given it's an atomic op and
we already have it locally when we check for hostns. Also pass in
svc directly into session affinity APIs to simplify them a bit and
move the multiple defined(ENABLE_SESSION_AFFINITY) && defined(BPF_HAVE_NETNS_COOKIE)
out of the bpf_sock.c code by letting compiler do most of the work
to optimize away the constants - we add *_{netns,addr}() api variants
and for the former we can place the BPF_HAVE_NETNS_COOKIE ifdef into
lb.h header to keep the code flow more readable in bpf_sock.c. At the
same time this also fixes couple of -Wdeclaration-after-statement
warnings.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: enable -nostdinc and other stricter compilation options

We want to have our code as close to usual kernel conventions as
possible to ease review, so add a few more switches for clang
where we emit useful warnings in our code if it's not the case,
that is, -std=gnu89 and -Wdeclaration-after-statement.

We also use quite a number of shadow declarations where they are
really useless; I'm not sure whether clang ends up allocating new
stack space for them; in any case lets add -Wshadow given the
current locations are fixed now.

Also, given our headers are all self-contained, add -nostdinc to
fully /guarantee/ that clang does not search the standard include
paths anymore.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: refactor and improve perf event output message init

Prep work and general refactoring for enabling stricter compilation
options closer to the kernel. In order to avoid warnings from the
-Wdeclaration-after-statement, make use of compound literals where
adequate.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: avoid potentially expensive calls into get_hash_recalc

Depending on the NICs hash, we might end up recalculating it
from the kernel flow dissector to get a higher quality L4 hash
on the skb instead of L3 one, for example.

We used to do this mainly for the skb->hash based service LB,
but nowadays all occurences in the code are for cilium monitor
debug output where it's used as a heuristic to see that the event
messages belong to the same flow.

Hubble side doesn't do anything with the hash.

Whether L3 or L4 hash doesn't really matter too much here, so
get rid of these calls.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/lib/policy_log.h:
```diff
@@ -47,30 +47,31 @@ send_policy_verdict_notify(struct __ctx_buff *ctx, __u32 remote_label, __u16 dst
+	msg.ipv6 = is_ipv6;
+
+	msg.pad0 = 0;
+	msg.pad1 = 0;
```
The nice thing about the previous initializer was that we get auto-zeroed fields if it's extended and a developer forgets to set the field, whereas now it's easier to break right?

joestringer commented on bpf/lib/trace.h:
```diff
@@ -151,29 +155,28 @@ send_trace_notify(struct __ctx_buff *ctx, __u8 obs_point, __u32 src, __u32 dst,
+	msg.type = CILIUM_NOTIFY_TRACE;
+	msg.subtype = obs_point;
+	msg.source = EVENT_SOURCE;
+	msg.version = NOTIFY_CAPTURE_VER;
```
can't help but wonder if there's some better code refactoring/sharing we could do when reviewing ~three~ four separate, very similar, initializers like this ;)

borkmann replied:
I'll make it a helper to simplify all the other types like this and address the other comment from above.

joestringer reviewed (approved):
Couple minor nits but changes LGTM.

pchaigno commented on bpf/Makefile.bpf:
```diff
@@ -3,11 +3,14 @@
 FLAGS := -I$(ROOT_DIR)/bpf/include -I$(ROOT_DIR)/bpf -D__NR_CPUS__=$(shell nproc) -O2 -g
 
-CLANG_FLAGS := ${FLAGS} -target bpf -emit-llvm
+CLANG_FLAGS := ${FLAGS} -target bpf -std=gnu89 -nostdinc -emit-llvm
```
Do we need the same flags in `test/bpf/Makefile`?

borkmann replied:
Yeah probably makes sense, will add, the CFLAGS lack also few other things in there.

qmonnet commented on bpf/lib/dbg.h:
```diff
@@ -148,34 +149,28 @@ struct debug_msg {
 static __always_inline void cilium_dbg(struct __ctx_buff *ctx, __u8 type,
 				       __u32 arg1, __u32 arg2)
 {
-	__u32 hash = get_hash_recalc(ctx);
```
All the `get_hash_recalc()` removal should be moved to first commit?

borkmann replied:
I moved it into the __notify_common_hdr() in the 2nd one, I'm also fine to remove the macro entirely, no strong opinion.

qmonnet replied:
Ah right it's moved, not removed. Ok it's fine as is then, sorry.

qmonnet reviewed (commented):
Looks like removal of `get_hash_recalc()` got split between the two patches.

Otherwise looks good, but second patch is huge. Did all of those fixes came out of the new warnings at compile time?

qmonnet commented on bpf/lib/trace.h:
```diff
@@ -143,112 +143,87 @@ static __always_inline void
 		   __u16 dst_id, __u32 ifindex, __u8 reason, __u32 monitor)
 {
+	__u64 ctx_len = ctx_full_len(ctx);
+	__u64 cap_len = min_t(__u64, !monitor ? TRACE_PAYLOAD_LEN : monitor,
```
Nit: `!monitor ? TRACE_PAYLOAD_LEN : monitor` is the same as `monitor ? : TRACE_PAYLOAD_LEN` I think, which might be more readable?

borkmann replied:
Ah yeah of course, will fix :-)

qmonnet commented on bpf/bpf_overlay.c:
```diff
@@ -113,7 +115,6 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 	if (1) {
```
Not related to that PR, but I've been curious: Do you know what those `if (1)` are for?

borkmann replied:
My assumption is to get a new scope for declaring vars, here host_mac and router_mac. Could also have been plain `{ ... }`to make it more clear.

qmonnet reviewed (approved):
Looks great!

borkmann committed:
bpf: improve overlay build permutation coverage

Given bpf_overlay is similarly serving BPF NodePort like bpf_netdev does,
improve coverage so we can easier spot any build issues.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
